README.md
# Hanami::Mailer

Mail for Ruby applications.

## Version

**This branch contains the code for `hanami-mailer` 2.x.**

## Status

[![Gem Version](https://badge.fury.io/rb/hanami-mailer.svg)](https://badge.fury.io/rb/hanami-mailer)
[![CI](https://github.com/hanami/mailer/workflows/ci/badge.svg?branch=main)](https://github.com/hanami/mailer/actions?query=workflow%3Aci+branch%3Amain)
[![Test Coverage](https://codecov.io/gh/hanami/mailer/branch/main/graph/badge.svg)](https://codecov.io/gh/hanami/mailer)
[![Depfu](https://badges.depfu.com/badges/739c6e10eaf20d3ba4240d00828284db/overview.svg)](https://depfu.com/github/hanami/mailer?project=Bundler)
[![Inline Docs](http://inch-ci.org/github/hanami/mailer.svg)](http://inch-ci.org/github/hanami/mailer)

## Contact

* Home page: http://hanamirb.org
* Mailing List: http://hanamirb.org/mailing-list
* API Doc: http://rdoc.info/gems/hanami-mailer
* Bugs/Issues: https://github.com/hanami/mailer/issues
* Support: http://stackoverflow.com/questions/tagged/hanami
* Chat: http://chat.hanamirb.org

## Rubies

__Hanami::Mailer__ supports Ruby (MRI) 2.6+

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'hanami-mailer'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install hanami-mailer

## Usage

### Conventions

  * Templates are searched under `Hanami::Mailer::Configuration#root`, set this value according to your app structure (eg. `"app/templates"`).
  * A mailer will look for a template with a file name that is composed by its full class name (eg. `"articles/index"`).
  * A template must have two concatenated extensions: one for the format and one for the engine (eg. `".html.erb"`).
  * The framework must be loaded before rendering the first time: `Hanami::Mailer.finalize(configuration)`.

### Mailers

A simple mailer looks like this:

```ruby
require 'hanami/mailer'
require 'ostruct'

# Create two files: `invoice.html.erb` and `invoice.txt.erb`

configuration = Hanami::Mailer::Configuration.new do |config|
  config.delivery_method = :test
end

class Invoice < Hanami::Mailer
  from "noreply@example.com"
  to ->(locals) { locals.fetch(:user).email }
end

configuration = Hanami::Mailer.finalize(configuration)

invoice = OpenStruct.new(number: 23)
mailer  = InvoiceMailer.new(configuration: configuration)
mail    = mailer.deliver(invoice: invoice)

mail
  # => #<Mail::Message:70303354246540, Multipart: true, Headers: <Date: Wed, 22 Mar 2017 11:48:57 +0100>, <From: noreply@example.com>, <To: user@example.com>, <Cc: >, <Bcc: >, <Message-ID: <58d25699e47f9_b4e13ff0c503e4f4632e6@escher.mail>>, <Subject: >, <Mime-Version: 1.0>, <Content-Type: multipart/alternative; boundary=--==_mimepart_58d25699e42d2_b4e13ff0c503e4f463186>, <Content-Transfer-Encoding: 7bit>>

mail.to_s
  # =>
  # From: noreply@example.com
  # To: user@example.com
  # Message-ID: <58d25699e47f9_b4e13ff0c503e4f4632e6@escher.mail>
  # Subject:
  # Mime-Version: 1.0
  # Content-Type: multipart/alternative;
  #  boundary="--==_mimepart_58d25699e42d2_b4e13ff0c503e4f463186";
  #  charset=UTF-8
  # Content-Transfer-Encoding: 7bit
  #
  #
  # ----==_mimepart_58d25699e42d2_b4e13ff0c503e4f463186
  # Content-Type: text/plain;
  #  charset=UTF-8
  # Content-Transfer-Encoding: 7bit
  #
  # Invoice #23
  #
  # ----==_mimepart_58d25699e42d2_b4e13ff0c503e4f463186
  # Content-Type: text/html;
  #  charset=UTF-8
  # Content-Transfer-Encoding: 7bit
  #
  # <html>
  #   <body>
  #     <h1>Invoice template</h1>
  #   </body>
  # </html>
  #
  # ----==_mimepart_58d25699e42d2_b4e13ff0c503e4f463186--
```

A mailer with `.to` and `.from` addresses and mailer delivery:

```ruby
require 'hanami/mailer'

configuration = Hanami::Mailer::Configuration.new do |config|
  config.delivery_method = :smtp,
                           address:              "smtp.gmail.com",
                           port:                 587,
                           domain:               "example.com",
                           user_name:            ENV['SMTP_USERNAME'],
                           password:             ENV['SMTP_PASSWORD'],
                           authentication:       "plain",
                           enable_starttls_auto: true
end

class WelcomeMailer < Hanami::Mailer
  return_path 'bounce@sender.com'
  from 'noreply@sender.com'
  to   'noreply@recipient.com'
  cc   'cc@sender.com'
  bcc  'alice@example.com'

  subject 'Welcome'
end

WelcomeMailer.new(configuration: configuration).call(locals)
```

### Locals

The set of objects passed in the `deliver` call are called `locals` and are available inside the mailer and the template.

```ruby
require 'hanami/mailer'
require 'ostruct'

user = OpenStruct.new(name: Luca', email: 'user@hanamirb.org')

class WelcomeMailer < Hanami::Mailer
  from    'noreply@sender.com'
  subject 'Welcome'
  to      ->(locals) { locals.fetch(:user).email }
end

WelcomeMailer.new(configuration: configuration).deliver(user: luca)
```

The corresponding `erb` file:

```erb
Hello <%= user.name %>!
```

### Scope

All public methods defined in the mailer are accessible from the template:

```ruby
require 'hanami/mailer'

class WelcomeMailer < Hanami::Mailer
  from    'noreply@sender.com'
  to      'noreply@recipient.com'
  subject 'Welcome'

  def greeting
    'Ahoy'
  end
end
```

```erb
<h2><%= greeting %></h2>
```

### Template

The template file must be located under the relevant `root` and must match the inflected snake case of the mailer class name.

```ruby
# Given this root
configuration.root      # => #<Pathname:app/templates>

# For InvoiceMailer, it looks for:
#  * app/templates/invoice_mailer.html.erb
#  * app/templates/invoice_mailer.txt.erb
```

If we want to specify a different template, we can do:

```ruby
class InvoiceMailer < Hanami::Mailer
  template 'invoice'
end

# It will look for:
#  * app/templates/invoice.html.erb
#  * app/templates/invoice.txt.erb
```

### Engines

The builtin rendering engine is [ERb](http://en.wikipedia.org/wiki/ERuby).

This is the list of the supported engines.
They are listed in order of **higher precedence**, for a given extension.
For instance, if [ERubis](http://www.kuwata-lab.com/erubis/) is loaded, it will be preferred over ERb to render `.erb` templates.

<table>
  <tr>
    <th>Engine</th>
    <th>Extensions</th>
  </tr>
  <tr>
    <td>Erubis</td>
    <td>erb, rhtml, erubis</td>
  </tr>
  <tr>
    <td>ERb</td>
    <td>erb, rhtml</td>
  </tr>
  <tr>
    <td>Redcarpet</td>
    <td>markdown, mkd, md</td>
  </tr>
  <tr>
    <td>RDiscount</td>
    <td>markdown, mkd, md</td>
  </tr>
  <tr>
    <td>Kramdown</td>
    <td>markdown, mkd, md</td>
  </tr>
  <tr>
    <td>Maruku</td>
    <td>markdown, mkd, md</td>
  </tr>
  <tr>
    <td>BlueCloth</td>
    <td>markdown, mkd, md</td>
  </tr>
  <tr>
    <td>Asciidoctor</td>
    <td>ad, adoc, asciidoc</td>
  </tr>
  <tr>
    <td>Builder</td>
    <td>builder</td>
  </tr>
  <tr>
    <td>CSV</td>
    <td>rcsv</td>
  </tr>
  <tr>
    <td>CoffeeScript</td>
    <td>coffee</td>
  </tr>
  <tr>
    <td>WikiCloth</td>
    <td>wiki, mediawiki, mw</td>
  </tr>
  <tr>
    <td>Creole</td>
    <td>wiki, creole</td>
  </tr>
  <tr>
    <td>Etanni</td>
    <td>etn, etanni</td>
  </tr>
  <tr>
    <td>Haml</td>
    <td>haml</td>
  </tr>
  <tr>
    <td>Less</td>
    <td>less</td>
  </tr>
  <tr>
    <td>Liquid</td>
    <td>liquid</td>
  </tr>
  <tr>
    <td>Markaby</td>
    <td>mab</td>
  </tr>
  <tr>
    <td>Nokogiri</td>
    <td>nokogiri</td>
  </tr>
  <tr>
    <td>Plain</td>
    <td>html</td>
  </tr>
  <tr>
    <td>RDoc</td>
    <td>rdoc</td>
  </tr>
  <tr>
    <td>Radius</td>
    <td>radius</td>
  </tr>
  <tr>
    <td>RedCloth</td>
    <td>textile</td>
  </tr>
  <tr>
    <td>Sass</td>
    <td>sass</td>
  </tr>
  <tr>
    <td>Scss</td>
    <td>scss</td>
  </tr>
  <tr>
    <td>Slim</td>
    <td>slim</td>
  </tr>
  <tr>
    <td>String</td>
    <td>str</td>
  </tr>
  <tr>
    <td>Yajl</td>
    <td>yajl</td>
  </tr>
</table>


### Configuration

__Hanami::Mailer__ can be configured with a DSL that determines its behavior.
It supports a few options:

```ruby
require "hanami/mailer"

configuration = Hanami::Mailer::Configuration.new do |config|
  # Set the root path where to search for templates
  # Argument: String, Pathname, #to_pathname, defaults to the current directory
  #
  config.root = "path/to/root"

  # Set the default charset for emails
  # Argument: String, defaults to "UTF-8"
  #
  config.default_charset = "iso-8859"

  # Set the delivery method
  # Argument: Symbol
  # Argument: Hash, optional configurations
  config.delivery_method = :stmp
end
```

### Attachments

Attachments can be added with the following API:

```ruby
class InvoiceMailer < Hanami::Mailer
  # ...
  before do |mail, locals|
    mail.attachments["invoice-#{locals.fetch(:invoice).number}.pdf"] = 'path/to/invoice.pdf'
  end
end
```

### Delivery Method

The global delivery method is defined through the __Hanami::Mailer__ configuration, as:

```ruby
configuration = Hanami::Mailer::Configuration.new do |config|
  config.delivery_method = :smtp
end
```

```ruby
configuration = Hanami::Mailer::Configuration.new do |config|
  config.delivery_method = :smtp, { address: "localhost", port: 1025 }
end
```

Builtin options are:

  * Exim (`:exim`)
  * Sendmail (`:sendmail`)
  * SMTP (`:smtp`, for local installations)
  * SMTP Connection (`:smtp_connection`, via `Net::SMTP` - for remote installations)
  * Test (`:test`, for testing purposes)

### Custom Delivery Method

Developers can specify their own custom delivery policy:

```ruby
require 'hanami/mailer'

class MandrillDeliveryMethod
  def initialize(options)
    @options = options
  end

  def deliver!(mail)
    # ...
  end
end

configuration = Hanami::Mailer::Configuration.new do |config|
  config.delivery_method = MandrillDeliveryMethod,
                           username: ENV['MANDRILL_USERNAME'],
                           password: ENV['MANDRILL_API_KEY']
end
```

The class passed to `.delivery_method=` must accept an optional set of options
with the constructor (`#initialize`) and respond to `#deliver!`.

### Multipart Delivery

All the email are sent as multipart messages by default.
For a given mailer, the framework looks up for associated text (`.txt`) and `HTML` (`.html`) templates and render them.

```ruby
InvoiceMailer.new(configuration: configuration).deliver({})           # delivers both text and html templates
InvoiceMailer.new(configuration: configuration).deliver(format: :txt) # delivers only text template
```

Please note that **they aren't both mandatory, but at least one of them MUST** be present.

## Versioning

__Hanami::Mailer__ uses [Semantic Versioning 2.0.0](http://semver.org)

## Copyright

Copyright © 2015-2021 Luca Guidi – Released under MIT License

This project was formerly known as Lotus (`lotus-mailer`).


CHANGELOG.md
# Hanami::Mailer
Mail for Ruby applications

## v2.0.0.alpha1 (unreleased)
### Changed
- [Luca Guidi] Drop support for Ruby: MRI 2.3, 2.4, 2.5, 2.6

## v1.3.3 - 2021-01-14
### Added
- [Luca Guidi] Official support for Ruby: MRI 3.0

## v1.3.2 - 2020-02-03
### Added
- [Luca Guidi] Official support for Ruby: MRI 2.7
- [glaszig] Added `Hanami::Mailer.return_path` and `#return_path` to specify `MAIL FROM` address

## v1.3.1 - 2019-01-18
### Added
- [Luca Guidi] Official support for Ruby: MRI 2.6
- [Luca Guidi] Support `bundler` 2.0+

## v1.3.0 - 2018-10-24
### Added
- [Ben Bachhuber] Added support for `reply_to`

## v1.3.0.beta1 - 2018-08-08
### Added
- [Luca Guidi] Official support for JRuby 9.2.0.0

## v1.2.0 - 2018-04-11

## v1.2.0.rc2 - 2018-04-06

## v1.2.0.rc1 - 2018-03-30

## v1.2.0.beta2 - 2018-03-23

## v1.2.0.beta1 - 2018-02-28
### Added
- [Luca Guidi] Official support for Ruby: MRI 2.5

## v1.1.0 - 2017-10-25

## v1.1.0.rc1 - 2017-10-16

## v1.1.0.beta3 - 2017-10-04

## v1.1.0.beta2 - 2017-10-03

## v1.1.0.beta1 - 2017-08-11

## v1.0.0 - 2017-04-06

## v1.0.0.rc1 - 2017-03-31
### Fixed
- [Luca Guidi] Let `Hanami::Mailer.deliver` to bubble up `ArgumentError` exceptions

## v1.0.0.beta2 - 2017-03-17

## v1.0.0.beta1 - 2017-02-14
### Added
- [Luca Guidi] Official support for Ruby: MRI 2.4

## v0.4.0 - 2016-11-15
### Changed
- [Luca Guidi] Official support for Ruby: MRI 2.3+ and JRuby 9.1.5.0+

## v0.3.0 - 2016-07-22
### Added
- [Anton Davydov] Blind carbon copy (bcc) option
- [Anton Davydov] Carbon copy (cc) option

### Changed
- [Luca Guidi] Drop support for Ruby 2.0 and 2.1

## v0.2.0 - 2016-01-22
### Changed
- [Luca Guidi] Renamed the project

## v0.1.0 - 2015-09-30
### Added
- [Ines Coelho & Rosa Faria & Luca Guidi] Email delivery
- [Ines Coelho & Rosa Faria & Luca Guidi] Attachments
- [Ines Coelho & Rosa Faria & Luca Guidi] Multipart rendering
- [Ines Coelho & Rosa Faria & Luca Guidi] Configuration
- [Ines Coelho & Rosa Faria & Luca Guidi] Official support for Ruby 2.0


lib/hanami/mailer.rb
# frozen_string_literal: true

require "mail"
require "concurrent"

# Hanami
#
# @since 0.1.0
module Hanami
  # Hanami::Mailer
  #
  # @since 0.1.0
  class Mailer
    require "hanami/mailer/version"
    require "hanami/mailer/template"
    require "hanami/mailer/finalizer"
    require "hanami/mailer/configuration"
    require "hanami/mailer/dsl"

    # Content types mapping
    #
    # @since 0.1.0
    # @api private
    CONTENT_TYPES = {
      html: "text/html",
      txt: "text/plain"
    }.freeze

    private_constant(:CONTENT_TYPES)

    # Base error for Hanami::Mailer
    #
    # @since 0.1.0
    class Error < ::StandardError
    end

    # Unknown mailer
    #
    # This error is raised at the runtime when trying to deliver a mail message,
    # by using a configuration that it wasn't finalized yet.
    #
    # @since next
    # @api unstable
    #
    # @see Hanami::Mailer.finalize
    class UnknownMailerError < Error
      # @param mailer [Hanami::Mailer] a mailer
      #
      # @since next
      # @api unstable
      def initialize(mailer)
        super("Unknown mailer: #{mailer.inspect}. Please finalize the configuration before to use it.")
      end
    end

    # Missing delivery data error
    #
    # It's raised when a mailer doesn't specify `from` or `to`.
    #
    # @since 0.1.0
    class MissingDeliveryDataError < Error
      def initialize
        super("Missing delivery data, please check 'from', or 'to'")
      end
    end

    # @since next
    # @api unstable
    @_subclasses = Concurrent::Array.new

    # Override Ruby's hook for modules.
    # It includes basic `Hanami::Mailer` modules to the given Class.
    # It sets a copy of the framework configuration
    #
    # @param base [Class] the target mailer
    #
    # @since next
    # @api unstable
    def self.inherited(base)
      super
      @_subclasses.push(base)
      base.extend Dsl
    end

    private_class_method :inherited

    # Finalize the configuration
    #
    # This should be used before to start to use the mailers
    #
    # @param configuration [Hanami::Mailer::Configuration] the configuration to
    #   finalize
    #
    # @return [Hanami::Mailer::Configuration] the finalized configuration
    #
    # @since next
    # @api unstable
    #
    # @example
    #   require 'hanami/mailer'
    #
    #   configuration = Hanami::Mailer::Configuration.new do |config|
    #     # ...
    #   end
    #
    #   configuration = Hanami::Mailer.finalize(configuration)
    #   MyMailer.new(configuration: configuration)
    def self.finalize(configuration)
      Finalizer.finalize(@_subclasses, configuration)
    end

    # Initialize a mailer
    #
    # @param configuration [Hanami::Mailer::Configuration] the configuration
    # @return [Hanami::Mailer]
    #
    # @since 0.1.0
    def initialize(configuration:)
      @configuration = configuration
      freeze
    end

    # Prepare the email message when a new mailer is initialized.
    #
    # @return [Mail::Message] the delivered email
    #
    # @since 0.1.0
    # @api unstable
    #
    # @see Hanami::Mailer::Configuration#default_charset
    #
    # @example
    #   require 'hanami/mailer'
    #
    #   configuration = Hanami::Mailer::Configuration.new do |config|
    #     config.delivery_method = :smtp
    #   end
    #
    #   configuration = Hanami::Mailer.finalize(configuration)
    #
    #   module Billing
    #     class InvoiceMailer < Hanami::Mailer
    #       from    'noreply@example.com'
    #       to      ->(locals) { locals.fetch(:user).email }
    #       subject ->(locals) { "Invoice number #{locals.fetch(:invoice).number}" }
    #
    #       before do |mail, locals|
    #         mail.attachments["invoice-#{locals.fetch(:invoice).number}.pdf"] =
    #           File.read('/path/to/invoice.pdf')
    #       end
    #     end
    #   end
    #
    #   invoice = Invoice.new(number: 23)
    #   user    = User.new(name: 'L', email: 'user@example.com')
    #
    #   mailer = Billing::InvoiceMailer.new(configuration: configuration)
    #
    #   # Deliver both text, HTML parts and the attachment
    #   mailer.deliver(invoice: invoice, user: user)
    #
    #   # Deliver only the text part and the attachment
    #   mailer.deliver(invoice: invoice, user: user, format: :txt)
    #
    #   # Deliver only the text part and the attachment
    #   mailer.deliver(invoice: invoice, user: user, format: :html)
    #
    #   # Deliver both the parts with "iso-8859"
    #   mailer.deliver(invoice: invoice, user: user, charset: 'iso-8859')
    def deliver(locals)
      mail(locals).deliver
    rescue ArgumentError => exception
      raise MissingDeliveryDataError if exception.message =~ /SMTP (From|To) address/

      raise
    end

    # @since next
    # @api unstable
    alias_method :call, :deliver

    # Render a single template with the specified format.
    #
    # @param format [Symbol] format
    #
    # @return [String] the output of the rendering process.
    #
    # @since 0.1.0
    # @api unstable
    def render(format, locals)
      template(format).render(self, locals)
    end

    private

    # @api unstable
    # @since next
    attr_reader :configuration

    # @api unstable
    # @since next
    def mail(locals)
      Mail.new.tap do |mail|
        instance_exec(mail, locals, &self.class.before)
        bind(mail, locals)
      end
    end

    # @api unstable
    # @since next
    #
    def bind(mail, locals) # rubocop:disable Metrics/AbcSize
      charset = locals.fetch(:charset, configuration.default_charset)

      mail.return_path = __dsl(:return_path, locals)
      mail.from        = __dsl(:from,        locals)
      mail.to          = __dsl(:to,          locals)
      mail.cc          = __dsl(:cc,          locals)
      mail.bcc         = __dsl(:bcc,         locals)
      mail.reply_to    = __dsl(:reply_to,    locals)
      mail.subject     = __dsl(:subject,     locals)

      mail.html_part = __part(:html, charset, locals)
      mail.text_part = __part(:txt,  charset, locals)

      mail.charset = charset
      mail.delivery_method(*configuration.delivery_method)
    end

    # @since next
    # @api unstable
    def template(format)
      configuration.template(self.class, format)
    end

    # @since 0.1.0
    # @api unstable
    def __dsl(method_name, locals)
      case result = self.class.__send__(method_name)
      when Proc
        result.call(locals)
      else
        result
      end
    end

    # @since 0.1.0
    # @api unstable
    def __part(format, charset, locals)
      return unless __part?(format, locals)

      Mail::Part.new.tap do |part|
        part.content_type = "#{CONTENT_TYPES.fetch(format)}; charset=#{charset}"
        part.body         = render(format, locals)
      end
    end

    # @since 0.1.0
    # @api unstable
    def __part?(format, locals)
      wanted = locals.fetch(:format, nil)
      wanted == format ||
        (!wanted && !template(format).nil?)
    end
  end
end
